### DISCUSSION TITLE
Create a strimzi CustomResource in a Golang - operator-sdk based operator

### DISCUSSION THREAD

**Discussion by DanArlowski** [strimzi/strimzi-kafka-operator#5095] [Q&A] [answered]
when creating a Golang based operator, to use an external custom resources you add

```
import (
pkgname "github.com/some/package/api/v1
)

...

utilruntime.Must(pkgname.AddToScheme(scheme))

```

I was wondering how to create a strimzi Cr using my operator, as the operator is not written natively in go, so the API package does not exist in go.

**Comment by ppatierno** [comment, marked as answer]
We currently support the API in Java of course that helps but not in Go.
I think the only way you have is kind of writing the api package by yourself. I mean, you have to write the corresponding Go structs mapping the Strimzi related custom resource. The simpler way would be having a tool converting the provided CRDs (YAML) into Go structs but I am not aware of such a tool.

**Reply by DanArlowski** [reply]
suggest supporting it officially as the api may can (and probably will) change,

as such - this can help most go-based k8s operators and apps.

nonetheless - thanks for you help and quick answer.

**Reply by scholzj** [reply]
> The simpler way would be having a tool converting the provided CRDs (YAML) into Go structs but I am not aware of such a tool.

I know Pulumi has a tool which is quite similar to this, but I think it generates the Golang structures for some Pulumi client and not the Kubernetes client: https://www.pulumi.com/blog/introducing-crd2pulumi/ ... not sure how easy it would be to change these or use them with the Kubernetes client.

> suggest supporting it officially as the api may can (and probably will) change,

The CRD generator (https://github.com/strimzi/strimzi-kafka-operator/tree/main/crd-generator) currently generates the docs and the CRD YAMLs. It would need to be extended to generate the Golang files. Contributions are always welcomed ;-).

---

Other option should IMHO be to use the dynamic Kubernetes client -> basically just use `map[string]interface{}` structures as described here: https://github.com/kubernetes/client-go/blob/master/examples/dynamic-create-update-delete-deployment/main.go ... then you should not need any generated types

**Comment by tusharm** [comment]
Hi! Just checking if things have changed since last comment 😄 . I have come across a similar usecase and would like to know if client-go `dynamic` pkg is still the way to go.

**Reply by scholzj** [reply]
No, I don't think anything changed since last time.

**Comment by scholzj** [comment]
FYI: I put together a generator for Golang APIs and Kubernetes clients: https://github.com/scholzj/strimzi-go ... if anyone is still interested in this, please check it out and let me know if you have any feedback and/or are interested in using it.
